src/Data/JSONPath/Types.hs

```diff
@@ -30,12 +33,32 @@ data SingularPathElement
   | Index Int
   deriving (Show, Eq)
 
+data FunctionExpr
+  = FunctionExpr FunctionName [FunctionArgument]
+  deriving (Show, Eq)
+
+data FunctionName
+  = FunLength
+  | FunCount
+  | FunMatch
+  | FunSearch
+  | FunValue
+  deriving (Show, Eq)
```

Comment:
I think it is better if these are be made into one type. This would allow catching arity errors at parse time and this type will not allow storing invalid function expressions.

Reply:
@akshaymankar I made `FunctionName` an alias for `Text`. Does that seem reasonable or do you have something better in mind?

Reply:
I was thinking something like:

```haskell
data FunctionExpr
  = Length FunctionArgument
  | Count FunctionArgument
  | ...
```